File: example/src/androidTest/java/eu/darken/mvpbakery/example/screens/counting/CountingFragmentTest.java
package eu.darken.mvpbakery.example.screens.counting;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import androidx.fragment.app.Fragment;
import dagger.android.AndroidInjector;
import eu.darken.mvpbakery.example.R;
import eu.darken.mvpbakery.injection.ManualInjector;
import testhelper.FragmentTestRule;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class CountingFragmentTest {
    @Rule public FragmentTestRule<CountingFragment> fragmentRule = new FragmentTestRule<>(CountingFragment.class);

    @Mock CountingPresenter presenter;

    private CountingComponent component = new CountingComponent() {
        @Override
        public CountingPresenter getPresenter() {
            return presenter;
        }

        @Override
        public void inject(CountingFragment overviewFragment) {

        }
    };

    private ManualInjector<Fragment> injector = new ManualInjector<Fragment>() {
        @Override
        public AndroidInjector get(Fragment fragment) {
            return component;
        }

        @Override
        public void inject(Fragment fragment) {

        }
    };

    @Before
    public void setUp() {
        doAnswer(invocationOnMock -> null).when(presenter).onBindChange(any());
        when(presenter.getComponent()).thenReturn(component);
        fragmentRule.setManualInjector(injector);
    }

    @After
    public void tearDown() {

    }

    @Test
    public void testShowText() throws Throwable {
        fragmentRule.launchActivity(null);
        onView(withId(R.id.fragment_button)).check(matches(withText("+1")));

        fragmentRule.runOnUiThread(() -> fragmentRule.getFragment().showText("Straw"));
        onView(withId(R.id.fragment_text)).check(matches(withText("Straw")));
        fragmentRule.runOnUiThread(() -> fragmentRule.getFragment().showText("berry"));
        onView(withId(R.id.fragment_text)).check(matches(withText("berry")));
    }

    @Test
    public void testCount() {
        fragmentRule.launchActivity(null);
        onView(withId(R.id.fragment_button)).check(matches(withText("+1")));

        onView(withId(R.id.fragment_button)).perform(click());
        verify(presenter).onCountClick();
    }

}

File: example/src/androidTest/java/eu/darken/mvpbakery/example/screens/counting/CountingFragmentTest.kt
package eu.darken.mvpbakery.example.screens.counting

import androidx.fragment.app.Fragment
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.espresso.matcher.ViewMatchers.withText
import dagger.android.AndroidInjector
import eu.darken.mvpbakery.example.R
import eu.darken.mvpbakery.injection.ManualInjector
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.verify
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import testhelper.FragmentTestRule

class CountingFragmentTest {
    @get:Rule
    var fragmentRule = FragmentTestRule(CountingFragment::class.java)

    @MockK
    lateinit var presenter: CountingPresenter

    private val component = object : CountingComponent {
        override val presenter: CountingPresenter
            get() = this@CountingFragmentTest.presenter

        override fun inject(overviewFragment: CountingFragment) = Unit
    }

    private val injector = object : ManualInjector<Fragment> {
        @Suppress("UNCHECKED_CAST")
        override fun get(instance: Fragment): AndroidInjector<Fragment> {
            return component as AndroidInjector<Fragment>
        }

        override fun inject(fragment: Fragment) = Unit
    }

    @Before
    fun setUp() {
        MockKAnnotations.init(this, relaxUnitFun = true)
        fragmentRule.manualInjector = injector
        every { presenter.component } returns component
    }

    @After
    fun tearDown() {

    }

    @Test
    @Throws(Throwable::class)
    fun testShowText() {
        fragmentRule.launchActivity(null)
        onView(withId(R.id.fragment_button)).check(matches(withText("+1")))

        fragmentRule.runOnUiThread { fragmentRule.fragment.showText("Straw") }
        onView(withId(R.id.fragment_text)).check(matches(withText("Straw")))
        fragmentRule.runOnUiThread { fragmentRule.fragment.showText("berry") }
        onView(withId(R.id.fragment_text)).check(matches(withText("berry")))
    }

    @Test
    fun testCount() {
        fragmentRule.launchActivity(null)
        onView(withId(R.id.fragment_button)).check(matches(withText("+1")))
        onView(withId(R.id.fragment_button)).perform(click())
        verify { presenter.onCountClick() }
    }

}
